### PR TITLE
feat: WitnessService

### PR DESCRIPTION
**Pull request by ukstv**
Create Witness CAR from Merkle CAR

### CODE REVIEW

**Review by stbrody — COMMENTED** [review]
looking really nice.  A few small questions, and main request is to increase the test coverage.

File: src/services/__tests__/fake-factory.util.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * Generate fake data for testing purposes.
+ */
+export class FakeFactory {
+  constructor(
+    readonly ipfsService: IIpfsService,
+    readonly metadataService: IMetadataService,
+    readonly requestRepository: RequestRepository
+  ) {}
+
+  async stream(): Promise<StreamID> {
```
[the remark this answers is not in the document]
Comment (ukstv) [reply src/services/__tests__/fake-factory.util.ts]: done.

File: src/services/witness-service.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  destination.blocks.put(source.blocks.get(cid) || raiseNoBlockError(cid, name))
+}
+
+export type WitnessCARVerification = {
+  anchorCommit: AnchorCommit
+  isCorrect: boolean
+}
+
+/**
+ * Extract Anchor Commit and verify if its Merkle path goes from Merkle root to the `.prev` commit.
```
[the remark this answers is not in the document]
Comment (ukstv) [reply src/services/witness-service.ts:38]: Okay, let it return boolean for now.

File: src/services/witness-service.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * @param witnessCAR - CAR file containing Merkle witness i.e. Anchor Commit, proof, Merkle root, and all the intermediary nodes.
+ */
+export function verifyWitnessCAR(witnessCAR: CAR): WitnessCARVerification {
```
[the remark this answers is not in the document]
Comment (ukstv) [reply src/services/witness-service.ts:42]: It depends on the actual usage. The function should eventually move to js-ceramic, and I am not sure now what would be the most convenient there.

File: src/services/witness-service.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  }
+  const isCorrect = currentCID.equals(anchorCommit.prev)
+  return {
+    anchorCommit: anchorCommit,
+    isCorrect: isCorrect,
+  }
+}
+
+export class WitnessService {
+  buildWitnessCAR(anchor: DatabaseAnchor, merkleCAR: CAR): CAR {
```
[the remark this answers is not in the document]
Comment (ukstv) [reply src/services/witness-service.ts:66]: good catch

File: src/services/__tests__/witness-service.test.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    tx,
+    merkleTree.root.data.cid
+  )
+  const anchors = await anchorService._createAnchorCommits(ipfsProofCid, merkleTree)
+  await anchorService._persistAnchorResult(anchors, candidates)
+
+  const witnessService = new WitnessService()
+  for (const freshAnchor of anchors) {
+    const anchor = await anchorRepository.findByRequestId(freshAnchor.requestId)
+    const witnessCAR = witnessService.buildWitnessCAR(anchor, merkleTree.car)
```
[the remark this answers is not in the document]
Comment (ukstv) [reply src/services/__tests__/witness-service.test.ts]: I understand that. I can't see the reason to have that kind of test. Could you please elaborate on why do you see that necessary?

[the remark this answers is not in the document]
Comment (stbrody) [reply src/services/__tests__/witness-service.test.ts]: To ensure we're not using more network bandwidth than is needed.  I could easily see bugs in the CAR construction logic causing us to spent a lot of resources building and sending CAR files that contain way more than is necessary.

File: src/services/witness-service.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  destination.blocks.put(source.blocks.get(cid) || raiseNoBlockError(cid, name))
+}
+
+export class InvalidWitnessCARError extends Error {
+  constructor(message: string) {
+    super(message)
+  }
+}
+
+export function* witnessCIDs(witnessCAR: CAR): Generator<CID> {
```
[the remark this answers is not in the document]
Comment (ukstv) [reply src/services/witness-service.ts:45]: Added!

**Pull request merged**